File: gin-enterprise-template-thin/internal/apiserver/biz/v1/menu/menu.go
package menu

import (
	"context"

	"github.com/clin211/gin-enterprise-template/internal/apiserver/store"
	v1 "github.com/clin211/gin-enterprise-template/pkg/api/apiserver/v1"
)

// MenuBiz 定义处理菜单请求所需的方法.
type MenuBiz interface {
	Create(ctx context.Context, rq *v1.CreateMenuRequest) (*v1.CreateMenuResponse, error)
	Update(ctx context.Context, rq *v1.UpdateMenuRequest) (*v1.UpdateMenuResponse, error)
	Delete(ctx context.Context, rq *v1.DeleteMenuRequest) (*v1.DeleteMenuResponse, error)
	Get(ctx context.Context, rq *v1.GetMenuRequest) (*v1.GetMenuResponse, error)
	List(ctx context.Context, rq *v1.ListMenuRequest) (*v1.ListMenuResponse, error)

	MenuExpansion
}

// MenuExpansion 定义菜单操作的扩展方法.
type MenuExpansion interface {
	// ListMenuTree 获取菜单树
	ListMenuTree(ctx context.Context, rq *v1.ListMenuTreeRequest) (*v1.ListMenuTreeResponse, error)
	// GetUserMenuTree 获取用户可见的菜单树
	GetUserMenuTree(ctx context.Context, rq *v1.GetUserMenuTreeRequest) (*v1.GetUserMenuTreeResponse, error)
}

// menuBiz 是 MenuBiz 接口的实现.
type menuBiz struct {
	store store.IStore
}

// 确保 menuBiz 实现了 MenuBiz 接口.
var _ MenuBiz = (*menuBiz)(nil)

func New(store store.IStore) *menuBiz {
	return &menuBiz{store: store}
}

File: web/src/service/api/menu.ts
import { request } from '../request';

/**
 * Get menu list
 *
 * @param params pageToken, pageSize, status, menuType, parentID
 */
export async function fetchGetMenuList(params?: Api.Menu.ListMenuRequest) {
  const result = await request<Api.Menu.ListMenuResponse>({
    url: '/v1/menus',
    params: {
      pageToken: params?.pageToken,
      pageSize: params?.pageSize,
      status: params?.status,
      menuType: params?.menuType,
      parentID: params?.parentID
    }
  });

  if (result.error || !result.data) {
    return result;
  }

  return {
    ...result,
    data: {
      totalCount: result.data.totalCount,
      menus: result.data.menus,
      pageToken: result.data.pageToken ?? ''
    } satisfies Api.Menu.ListMenuResponse
  };
}

/**
 * Create menu
 *
 * @param data Menu data
 */
export function fetchCreateMenu(data: Api.Menu.CreateMenuRequest) {
  return request<Api.Menu.CreateMenuResponse>({
    url: '/v1/menus',
    method: 'post',
    data
  });
}

/**
 * Get all menu tree (admin)
 *
 * @param params status filter
 */
export function fetchGetAllMenuTree(params?: { status?: number }) {
  return request<Api.Menu.ListMenuTreeResponse>({
    url: '/v1/menus/tree',
    params
  });
}

/**
 * Get menu by ID
 *
 * @param menuId Menu ID
 */
export function fetchGetMenu(menuId: string) {
  return request<Api.Menu.GetMenuResponse>({ url: `/v1/menus/${menuId}` });
}

/**
 * Update menu
 *
 * @param menuId Menu ID
 * @param data Update data
 */
export function fetchUpdateMenu(menuId: string, data: Api.Menu.UpdateMenuRequest) {
  return request<Api.Menu.UpdateMenuResponse>({
    url: `/v1/menus/${menuId}`,
    method: 'put',
    data
  });
}

/**
 * Delete menu
 *
 * @param menuId Menu ID
 */
export function fetchDeleteMenu(menuId: string) {
  return request<Api.Menu.DeleteMenuResponse>({
    url: `/v1/menus/${menuId}`,
    method: 'delete'
  });
}

/**
 * Get menu allowed roles
 *
 * @param menuId Menu ID
 */
export async function fetchGetMenuRoles(menuId: string) {
  const result = await request<Api.Menu.GetMenuRolesResponse>({
    url: `/v1/menus/${menuId}/roles`,
    method: 'get'
  });

  return result;
}

/**
 * Add role to menu
 *
 * @param menuId Menu ID
 * @param data Role data
 */
export async function fetchAddMenuRole(menuId: string, data: Api.Menu.AddMenuRoleRequest) {
  const result = await request({
    url: `/v1/menus/${menuId}/roles`,
    method: 'post',
    data
  });

  return result;
}

/**
 * Set menu roles (batch, overwrite mode)
 *
 * @param menuId Menu ID
 * @param data Role IDs
 */
export async function fetchSetMenuRoles(menuId: string, data: Api.Menu.SetMenuRolesRequest) {
  const result = await request<Api.Menu.SetMenuRolesResponse>({
    url: `/v1/menus/${menuId}/roles`,
    method: 'put',
    data
  });

  return result;
}

/**
 * Remove menu allowed role
 *
 * @param menuId Menu ID
 * @param roleId Role ID to remove
 */
export async function fetchRemoveMenuRole(menuId: string, roleId: string) {
  const result = await request({
    url: `/v1/menus/${menuId}/roles/${roleId}`,
    method: 'delete'
  });

  return result;
}
